File: src/_archived_pdf_parsing/pythonParsers/type3_parser.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Type3 PDF Parser - Qisqacha forma
Uses pdfplumber for accurate PDF text extraction
"""

import sys
import json
import pdfplumber
import re
from typing import Dict, List, Any


def clean(text: str = "") -> str:
    """Clean and normalize text"""
    return re.sub(r'\s+', ' ', str(text)).strip()


def extract(text: str, pattern: str) -> str:
    """Extract text using regex pattern"""
    match = re.search(pattern, text, re.IGNORECASE | re.DOTALL)
    return clean(match.group(1)) if match else ""


def between(text: str, start: str, end: str) -> str:
    """Extract text between two markers"""
    start_idx = text.find(start)
    if start_idx == -1:
        return ""

    from_idx = start_idx + len(start)
    end_idx = text.find(end, from_idx) if end else -1

    if end_idx == -1:
        end_idx = len(text)

    return clean(text[from_idx:end_idx])


# Field definitions for Type3
FIELD_DEFS = [
    {"label": "№", "pattern": r'№\s+(\d+)'},
    {"label": "Birlamchi tashxis", "pattern": r'1\.\s*Birlamchi tashxis\s+([^\n]+)'},
    {"label": "Bemor (mahalliy / kelgan)", "pattern": r'3\.\s*Bemor:\s+([A-Za-zА-Яа-я\s]+)'},
    {"label": "F.I.SH", "pattern": r'4\.\s*F\.I\.SH\s+(.+?)(?:\n|$)'},
    {"label": "Jinsi", "pattern": r'5\.\s*Jinsi\s+([A-Za-zА-Яа-я]+)'},
    {"label": "Tug'ilgan sanasi", "pattern": r'6\.\s*Tug\'ilgan sanasi\s+([0-9\-]+(?:\s+yil)?)'},
    {"label": "Doimiy yashash joyi", "pattern": r'7\.\s*Doimiy yashash joyi:\s*([^\n]+)'},
    {"label": "Hozirgi yashash joyi", "pattern": r'8\.\s*Hozirgi yashash joyi:\s*([^\n]+)'},
    {"label": "Ishlash / O'qish joyi", "pattern": r'9\.\s*Ishlash, o\'qish joyi.*?\s*([^\n]+)'},
    {"label": "Turar joyi bo'yicha davolash muassasasi", "pattern": r'11\.\s*Turar joyi bo\'yicha davolash muassasasi:\s*([^\n]+)'},
    {"label": "Bemor haqida habar olingan sana", "pattern": r'12\.\s*Bemor haqida habar.*?\s([0-9:\-\s]+)'},
    {"label": "Kim tomonidan yuborilgan", "pattern": r'13\.\s*Kim tomonidan yuborilgan.*?\s([^\n]+)'},
    {"label": "Kasallikning birinchi belgilari", "pattern": r'14\.\s*Kasallikning birinchi.*?belgilari\s+([^\n]+)'},
    {"label": "Bemor aniqlandi", "pattern": r'15\.\s*Bemor aniqlandi\s+([^\n]+)'},
    {"label": "Epidemiologik tekshiruv vaqti", "pattern": r'16\.\s*Epidemiologik tekshiruv.*?\s([0-9:\-\s]+)'},
    {"label": "Kasallangan sana", "pattern": r'17\.\s*Kasallangan.*?\s([0-9\-:\s]+)'},
    {"label": "Murojaat etgan vaqt", "pattern": r'18\.\s*Murojaat etgan vaqt.*?\s([0-9\-:\s]+)'},
    {"label": "Shoshilinch xabarnoma sanasi", "pattern": r'19\.\s*Shoshilinch xabarnoma.*?\s([0-9\-:\s]+)'},
    {"label": "Kasalxonaga yotqizilgan sana", "pattern": r'20\.\s*Kasalxonaga yotqizilgan.*?\s([0-9\-]+)'},
    {"label": "Yakunlovchi tashxis sanasi", "pattern": r'21\.\s*Yakunlovchi tashxis.*?\s([0-9\-]+)'},
    {"label": "Yotqizilgan kasalxona", "pattern": r'22\.\s*Yotqizilgan kasalxona.*?\s([^\n]+)'},
    {"label": "Uyida qoldirilish sababi", "pattern": r'23\.\s*Uyida qoldirilish.*?\s([^\n]+)'},
    {"label": "Laboratoriya tekshiruvi", "pattern": r'25\.\s*Laboratoriya tekshiruvi:\s*([^\n]+)'},
    {"label": "Oxirgi emlash haqida ma'lumot", "pattern": r'26\.\s*Oxirgi emlash.*?\s([^\n]+)'},
    {"label": "Epidemiolog vrach", "pattern": r'Epidemiolog vrach\s+([^\n]+)'},
    {"label": "Kartani topshirilgan vaqt", "pattern": r'Kartani.*?topshirilgan vaqt\s*([0-9\-]+)'},
]


def parse_pdf(pdf_path: str) -> Dict[str, Any]:
    """Parse Type3 PDF and extract all fields"""

    try:
        with pdfplumber.open(pdf_path) as pdf:
            # Extract text from all pages
            full_text = ""
            for page in pdf.pages:
                full_text += page.extract_text() or ""

            # Extract all fields
            parsed_fields = []

            for field_def in FIELD_DEFS:
                if 'pattern' in field_def:
                    value = extract(full_text, field_def['pattern'])
                elif 'start' in field_def and 'end' in field_def:
                    value = between(full_text, field_def['start'], field_def['end'])
                else:
                    value = ""

                parsed_fields.append({
                    "label": field_def['label'],
                    "value": value
                })

            return {
                "success": True,
                "parsed": parsed_fields,
                "raw": full_text,
                "metadata": {
                    "numPages": len(pdf.pages),
                    "info": pdf.metadata
                }
            }

    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "parsed": [],
            "raw": "",
            "metadata": {}
        }


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"success": False, "error": "PDF file path required"}))
        sys.exit(1)

    pdf_path = sys.argv[1]
    result = parse_pdf(pdf_path)
    print(json.dumps(result, ensure_ascii=False, indent=2))

File: src/_archived_pdf_parsing/pdfParsers/type3Parser.js
const pdf = require('pdf-parse');
const fs = require('fs').promises;

/**
 * Type 3 PDF Parser: Yuqumli Kasalliklar Kartasi (Qisqacha)
 * Bu parser 3.pdf formatidagi PDF fayllarni parse qiladi
 */

/** --- Universal Helper Functions --- */
function clean(str = "") {
  return String(str).replace(/\s+/g, " ").trim();
}

function extract(text, regex) {
  const m = text.match(regex);
  return m ? clean(m[1]) : "";
}

function between(text, start, end) {
  const i = text.indexOf(start);
  if (i === -1) return "";
  const from = i + start.length;

  let j = end ? text.indexOf(end, from) : -1;
  if (j === -1) j = text.length;

  return clean(text.substring(from, j));
}

/** --- 17–21 Sana blokini avtomatik topuvchi funksiyalar --- */
function extractTimeline(text) {
  const block = between(text, "17. Kasallangan", "22. Yotqizilgan kasalxona");

  const dates = block.match(/\d{2}-\d{2}-\d{4}(?:\s+\d{2}:\d{2})?/g) || [];

  return {
    kasallangan_sana: dates[0] || "",
    murojaat_vaqti: dates[1] || "",
    shoshilinch_xabarnoma: dates[2] || "",
    yotqizilgan_sana: dates[3] || "",
    yakunlovchi_sana: dates[4] || "",
  };
}

/** --- 3.pdf field definitions --- */
const FIELD_DEFS = [
  { label: "№", getValue: (txt) => extract(txt, /№\s+(\d+)/) },

  {
    label: "Birlamchi tashxis",
    getValue: (txt) => extract(txt, /1\. Birlamchi tashxis\s+([^\n]+)/),
  },

  {
    label: "Kasallik qo'zg'atuvchisi",
    getValue: (txt) =>
      between(txt, "2. Kasallik qo'zg'atuvchisining turi, xili", "3. Bemor"),
  },

  {
    label: "Bemor (mahalliy / kelgan)",
    getValue: (txt) => extract(txt, /3\. Bemor:\s+([A-Za-zА-Яа-я '' ]+)/),
  },

  {
    label: "F.I.SH",
    getValue: (txt) => extract(txt, /4\. F\.I\.SH\s+(.+)\n5\./),
  },

  {
    label: "Jinsi",
    getValue: (txt) => extract(txt, /5\. Jinsi\s+([A-Za-zА-Яа-я]+)/),
  },

  {
    label: "Tug'ilgan sanasi",
    getValue: (txt) => extract(txt, /6\. Tug'ilgan sanasi\s+([0-9\-]+ yil)/),
  },

  {
    label: "Doimiy yashash joyi",
    getValue: (txt) => {
      const m = txt.match(/7\. Doimiy yashash joyi:\s*([^\n]+)\n([^\n]+)/);
      return m ? clean(m[1] + " " + m[2]) : "";
    },
  },

  {
    label: "Hozirgi yashash joyi",
    getValue: (txt) => {
      const m = txt.match(/8\. Hozirgi yashash joyi:\s*([^\n]+)\n([^\n]+)/);
      return m ? clean(m[1] + " " + m[2]) : "";
    },
  },

  {
    label: "Ishlash / O'qish joyi",
    getValue: (txt) =>
      extract(txt, /9\. Ishlash, o'qish joyi, bog'cha\s*([^\n]*)/),
  },

  {
    label: "Turar joyi bo'yicha davolash muassasasi",
    getValue: (txt) =>
      extract(txt, /11\. Turar joyi bo'yicha davolash muassasasi:\s*([^\n]+)/),
  },

  {
    label: "Bemor haqida habar olingan sana",
    getValue: (txt) =>
      extract(txt, /12\. Bemor haqida habar olinganligi.*?\s([0-9:\- ]+)/),
  },

  {
    label: "Kim tomonidan yuborilgan",
    getValue: (txt) =>
      extract(txt, /13\. Kim tomonidan yuborilgan.*?\s([^\n]+)/),
  },

  {
    label: "Kasallikning birinchi belgilari",
    getValue: (txt) =>
      extract(
        txt,
        /14\. Kasallikning birinchi kunlaridagi asosiy belgilari\s+([^\n]+)/
      ),
  },

  {
    label: "Bemor aniqlandi",
    getValue: (txt) => extract(txt, /15\. Bemor aniqlandi\s+([^\n]+)/),
  },

  {
    label: "Epidemiologik tekshiruv vaqti",
    getValue: (txt) =>
      extract(txt, /16\. Epidemiologik tekshiruv.*?\s([0-9:\- ]+)/),
  },

  {
    label: "Kuzatuv tamomlangan",
    getValue: (txt) => extract(txt, /Kuzatuv tamomlangan vaqt\s+([0-9:\- ]+)/),
  },

  /** Timeline values from 17–21 */
  {
    label: "Kasallangan sana",
    getValue: (t) => extractTimeline(t).kasallangan_sana,
  },
  {
    label: "Murojaat etgan vaqt",
    getValue: (t) => extractTimeline(t).murojaat_vaqti,
  },
  {
    label: "Shoshilinch xabarnoma sanasi",
    getValue: (t) => extractTimeline(t).shoshilinch_xabarnoma,
  },
  {
    label: "Kasalxonaga yotqizilgan sana",
    getValue: (t) => extractTimeline(t).yotqizilgan_sana,
  },
  {
    label: "Yakunlovchi tashxis sanasi",
    getValue: (t) => extractTimeline(t).yakunlovchi_sana,
  },

  {
    label: "Yotqizilgan kasalxona va transport",
    getValue: (txt) => extract(txt, /22\. Yotqizilgan kasalxona.*?\s([^\n]+)/),
  },

  {
    label: "Uyida qoldirilish sababi",
    getValue: (txt) => extract(txt, /23\. Uyida qoldirilish.*?\s([^\n]+)/),
  },

  {
    label: "Laboratoriya tekshiruvi",
    getValue: (txt) => extract(txt, /25\. Laboratoriya tekshiruvi:\s*([^\n]+)/),
  },

  {
    label: "Oxirgi emlash haqida ma'lumot",
    getValue: (txt) => between(txt, "26. Oxirgi emlash", "II. KASALLIKNI"),
  },

  {
    label: "Taxminiy yuqqan muddati",
    getValue: (txt) =>
      extract(
        txt,
        /27\. Taxminiy yuqqan muddati:\s*([0-9\- ]+dan [0-9\- ]+gacha)/
      ),
  },

  {
    label: "Noqulay shart–sharoitlar",
    getValue: (txt) =>
      between(txt, "28. Kasallik yuqish", "29. Kasallik manbai"),
  },

  {
    label: "Kasallik manbai bo'lishi mumkin shaxslar",
    getValue: (txt) => between(txt, "29. Kasallik manbai", "30. Bu kasallikni"),
  },

  {
    label: "Suv va oziq–ovqat ma'lumotlari",
    getValue: (txt) =>
      between(txt, "30. Bu kasallikni kelib chiqishida", "A. YASHASH JOYI"),
  },

  {
    label: "Yashash sharoiti",
    getValue: (txt) => extract(txt, /31\. Yashash sharoiti\s+([^\n]+)/),
  },

  {
    label: "Zichligi – Kishilar",
    getValue: (txt) => extract(txt, /Kishilar soni\s+(\d+)/),
  },
  {
    label: "Zichligi – Xonalar",
    getValue: (txt) => extract(txt, /Xonalar soni\s+(\d+)/),
  },
  {
    label: "Zichligi – Maydon",
    getValue: (txt) => extract(txt, /Maydoni\s+(\d+)/),
  },

  {
    label: "Suv ta'minoti",
    getValue: (txt) => extract(txt, /33\. Suv bilan ta'minlanishi\s+([^\n]+)/),
  },

  {
    label: "Suyuq chiqindilar",
    getValue: (txt) => extract(txt, /34\. Suyuq chiqindilardan.*?\s([^\n]+)/),
  },

  {
    label: "Qattiq chiqindilar",
    getValue: (txt) => extract(txt, /35\. Qattiq chiqindilarni.*?\s([^\n]+)/),
  },

  {
    label: "Sanitariya holati",
    getValue: (txt) =>
      extract(txt, /36\. Xonadonda sanitar xolatni saqlash:\s*([^\n]+)/),
  },

  {
    label: "Bitlash borligi",
    getValue: (txt) => extract(txt, /37\. Bitlash borligi\s+([^\n]+)/),
  },

  {
    label: "Boshqa hashoratlar",
    getValue: (txt) => extract(txt, /Boshqa hashoratlar\s+([^\n]+)/),
  },

  {
    label: "Kemiruvchilar",
    getValue: (txt) => extract(txt, /Kemiruvchilar\s+([^\n]+)/),
  },

  {
    label: "Kasallik sabab (A bo'lim)",
    getValue: (txt) =>
      between(txt, "38. Kasallikni kelib chiqishida", "B. ISH"),
  },

  {
    label: "Bemor bo'lgan ob'ektlar",
    getValue: (txt) =>
      between(txt, "39. Bemor bo'lgan ob'ektlar", "40. Sanitar"),
  },

  {
    label: "Sanitar–gigienik talablar",
    getValue: (txt) => between(txt, "40. Sanitar – gigienik", "41. Kasallikni"),
  },

  {
    label: "Kasallikni keltirgan omillar",
    getValue: (txt) =>
      extract(txt, /41\. Kasallikni keltirib chiqar gan omillar\s+([^\n]+)/),
  },

  {
    label: "Tashqi muhit laboratorik tekshiruv",
    getValue: (txt) =>
      between(txt, "42. Tashqi muhit materiallarini", "III. O'СHOQ"),
  },

  {
    label: "Muloqatda bo'lganlar kuzatuvi",
    getValue: (txt) =>
      between(txt, "43. Bemor bilan muloqatda bo'lganlar", "44. Mahsus"),
  },

  {
    label: "Mahsus profilaktika tadbirlari",
    getValue: (txt) =>
      between(txt, "44. Mahsus profilaktik chora", "46. Kasallik o'chog'ida"),
  },

  {
    label: "Kasallanish mexanizmini uzish choralar",
    getValue: (txt) =>
      between(txt, "46. Kasallik o'chog'ida", "47. Bemor yotqizilgan"),
  },

  {
    label: "Bemor yotqizilgan kasalxona nomi",
    getValue: (txt) =>
      extract(txt, /47\. Bemor yotqizilgan kasalxona nomi\s*([^\n]+)/),
  },

  {
    label: "Kasallik chaqiruvchisini yuqtirgan joy",
    getValue: (txt) =>
      extract(
        txt,
        /1\. Kasallik chaqiruvchisini yuqtirishga gumon qilingan joy\s*([^\n]+)/
      ),
  },

  {
    label: "Yuqtirilishi ehtimol qiligan joy",
    getValue: (txt) =>
      extract(
        txt,
        /2\. Kasallikni yuqtirilishi ehtimol qilingan joy\s*([^\n]+)/
      ),
  },

  {
    label: "Gumon qilingan kasallik manbai",
    getValue: (txt) =>
      between(
        txt,
        "3. Gumon qilingan kasallik manbai",
        "4. Kasallik chaqiruvchisini"
      ),
  },

  {
    label: "Ehtimoliy asosiy faktor",
    getValue: (txt) =>
      between(txt, "4. Kasallik chaqiruvchisini", "5. Kasallikni yuqtirishga"),
  },

  {
    label: "Kasallikni yuqtirish sabablari",
    getValue: (txt) =>
      between(txt, "5. Kasallikni yuqtirishga sabab", "6. O'chog'dagi"),
  },

  {
    label: "O'choqdagi kasallanish",
    getValue: (txt) => extract(txt, /A:\s*([A-Za-z]+)/),
  },

  {
    label: "Kartani topshirilgan vaqt",
    getValue: (txt) =>
      extract(txt, /Kartani medstatistikka topshirilgan vaqt\s*([0-9\-]+)/),
  },

  {
    label: "Epidemiolog vrach",
    getValue: (txt) => extract(txt, /Epidemiolog vrach\s+([^\n]+)/),
  },
];

/** --- Main extractor --- */
function extractLabelValues(text) {
  return FIELD_DEFS.map((f) => ({
    label: f.label,
    value: f.getValue(text),
  }));
}

class Type3PDFParser {
  /**
   * PDF faylni parse qilish
   * @param {string} pdfPath - PDF fayl yo'li yoki Buffer
   * @returns {Promise<Object>} - Parse qilingan ma'lumotlar
   */
  static async parse(pdfPath) {
    try {
      let dataBuffer;
      if (Buffer.isBuffer(pdfPath)) {
        dataBuffer = pdfPath;
      } else {
        dataBuffer = await fs.readFile(pdfPath);
      }

      const data = await pdf(dataBuffer);
      const text = data.text || "";

      const result = {
        raw: text,
        parsed: extractLabelValues(text),
        metadata: {
          numPages: data.numpages,
          info: data.info,
        }
      };

      return result;
    } catch (error) {
      throw new Error(`Type3 PDF parsing error: ${error.message}`);
    }
  }

  /**
   * Sync variant - buffer uchun
   */
  static parseSync(text) {
    return extractLabelValues(text);
  }
}

module.exports = Type3PDFParser;
